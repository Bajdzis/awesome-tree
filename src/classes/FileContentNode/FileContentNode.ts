
export class FileContentNode {
    protected body:FileContentNode[];
    protected end:string;
    protected start:string;

    constructor(start?:string, end?:string) {
        this.start = start || '';
        this.body = [];
        this.end = end || '';
    }

    addChildren(children: FileContentNode){
        this.body.push(children);
    }

    getStart(){
        return this.start;
    }

    getEnd(){
        return this.end;
    }

    getChildren() {
        return [...this.body];
    }

    setChildren(children: FileContentNode[]){
        const newNode = new FileContentNode(this.start, this.end);
        newNode.body = children;
        return newNode;
    }

    getContent(): string{
        return `${this.start}${this.body.map(node => node.getContent()).join('')}${this.end}`;
    }

    isSame(toCompare:FileContentNode){
        return this.start === toCompare.start && this.end === toCompare.end;
    }

}
