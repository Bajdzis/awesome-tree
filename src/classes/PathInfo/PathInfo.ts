import { splitStringWithSplitter } from '../../utils/splitStringWithSplitter/splitStringWithSplitter';
import { WordsInfo } from '../WordsInfo/WordsInfo';



export class PathInfo {

    private path: string;
    private parts: WordsInfo[];
    private subParts: (WordsInfo|string)[][];
    private type: 'file' | 'directory';

    constructor(path:string) {
        this.path = path;
        this.type = path.match(/[\\/]+$/i) ? 'directory': 'file';
        this.subParts = splitStringWithSplitter(path,'\\/.')
            .map(part => {

                const {parts,isWord} = WordsInfo.splitByWord(part);

                return parts.reduce<(WordsInfo|string)[]>((acc, part) => {
                    if(isWord(part)) {
                        acc.push(new WordsInfo(part));
                    }else {
                        acc.push(part);
                    }
                    return acc;
                }, []);
            });

        this.parts =this.subParts.reduce<(WordsInfo)[]>((acc, parts) => {
            parts.forEach(part => {
                if(part instanceof WordsInfo) {
                    acc.push(part);
                }
            });
            return acc;
        }, []);
    }

    isFile() {
        return this.type === 'file';
    }

    isDirectory() {
        return this.type === 'directory';
    }

    getExtension(){
        if(this.isDirectory()){
            return '/DIRECTORY/';
        }
        return (/.([a-z]+)$/i).exec(this.path)?.[1];
    }

    getParent() {
        if (this.isFile()) {
            const directory = this.path.replace(/[^/\\]+$/, '');
            return new PathInfo(directory);
        }
        const parentDirectory = this.path.replace(/[^/\\]+(\/|\\)$/, '');
        return new PathInfo(parentDirectory);
    }

    getName(){
        const result = this.path.match( this.isFile() ?  /([^/\\]+)$/ : /([^/\\]+)[/\\]{1,1}$/);
        if(result) {
            return result[1];
        }

        return '';
    }

    getPath() {
        return this.path;
    }

    getParts() {
        return [...this.parts];
    }

    isSameNumberOfParts(instanceToCompare: PathInfo){
        return instanceToCompare.subParts.length === this.subParts.length;
    }

    isSimilarWords(instanceToCompare: PathInfo){
        if(!this.isSameNumberOfParts(instanceToCompare)) {
            return false;
        }
        return this.parts.every((parts, index) => {
            return parts.isSimilarly(instanceToCompare.getParts()[index]);
        });
    }

    isSimilar(instanceToCompare: PathInfo){
        return this.isSimilarWords(instanceToCompare) && this.getExtension() === instanceToCompare.getExtension();
    }

    getOnlySimilarPath(paths: PathInfo[]){
        return paths.filter(path => this.isSimilar(path));
    }

    includes(partPath: PathInfo) {
        if(partPath.isFile() || partPath.getParts().length > this.getParts().length){
            return false;
        }
        return partPath.getParts().every((part,index) => part.isSame(this.getParts()[index]));
    }

    includesSimilarly(partPath: PathInfo) {
        if(partPath.isFile() || partPath.getParts().length > this.getParts().length){
            return false;
        }
        return partPath.getParts().every((part,index) => part.isSimilarly(this.getParts()[index]));
    }

    // delete subParts in snapshot testing
    toJSON() {
        return {
            name: 'PathInfo',
            path: this.path,
            parts: this.parts,
            type: this.type,
        };
    }

}
