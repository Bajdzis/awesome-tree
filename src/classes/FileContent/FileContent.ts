import { FileContentNode } from '../FileContentNode/FileContentNode';
import { PathInfo } from '../PathInfo/PathInfo';

interface LineInfo {
    line: string;
    whiteSpaces:  number;
}
export class FileContent {
    private path: PathInfo;
    private content: string;

    constructor(path: PathInfo, content: string) {
        if(path.isDirectory()){
            throw new Error('[FileContent] Pass directory not file !');
        }
        this.path = path;
        this.content = content;
    }

    getContent(){
        return this.content;
    }

    getPath(){
        return this.path;
    }

    getExtension(){
        return this.path.getExtension();
    }

    getFileGraph(): FileContentNode {
        const file = new FileContentNode();
        const lines = this.content.split('\n');

        const countWhitespace = lines.map<LineInfo>((line, index) => {

            const extendString = index === lines.length -1 ? line : `${line}\n`;

            if (line.trim().length === 0) {
                return {
                    line: extendString,
                    whiteSpaces: Math.max(
                        FileContent.countNumberOfWhitespace(lines?.[index+1] || ''),
                        FileContent.countNumberOfWhitespace(lines?.[index-1] || '')
                    )
                };
            }

            return {
                line: extendString,
                whiteSpaces: FileContent.countNumberOfWhitespace(line)
            };
        });

        const children = this.nodeLine(countWhitespace);

        return file.setChildren(children);
    }

    nodeLine(lines:LineInfo[]): FileContentNode[]{
        if (lines.length === 0 ) {
            return [];
        }
        if (lines.length === 1 ) {
            return [new FileContentNode(lines[0].line)];
        }
        const baseWhitespace = lines[0].whiteSpaces;
        let lastIndexWithBaseSpace = 0;
        let omitElementsFromEnd = 0;
        const result: FileContentNode[] = [];

        for (let i = 1; i < lines.length; i++) {
            const prevLine = lines[i-1];
            const line = lines[i];


            if(line.whiteSpaces === baseWhitespace) {
                if(prevLine.whiteSpaces === line.whiteSpaces) {
                    const node = new FileContentNode(prevLine.line);
                    result.push(node);
                    omitElementsFromEnd = 1;
                } else {
                    const start = lines[lastIndexWithBaseSpace];
                    const node = new FileContentNode(start.line, line.line);
                    const childrenNodes = this.nodeLine(lines.slice(lastIndexWithBaseSpace + 1, i));
                    result.push(node.setChildren(childrenNodes));
                    omitElementsFromEnd = 1;
                    i++;
                }
                lastIndexWithBaseSpace = i;
            } else {
                omitElementsFromEnd++;
            }
        }

        if(omitElementsFromEnd!==0) {
            const omitLines = lines.slice(lines.length - omitElementsFromEnd, lines.length);
            omitLines.forEach(line => {
                result.push(new FileContentNode(line.line));
            });
        }

        return result;
    }

    static countNumberOfWhitespace(s0: string){
        for (let i = 0; i < s0.length; i++) {
            const letter = s0[i];
            if(letter === '\t' || letter === ' ') {
                continue;
            }
            return i;
        }

        return 0;
    }

}
