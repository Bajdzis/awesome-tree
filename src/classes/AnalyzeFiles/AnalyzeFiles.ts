import { FileContent } from '../FileContent/FileContent';
import { FileContentNodeAnalyze } from '../FileContentNodeAnalyze/FileContentNodeAnalyze';
import { PathInfo } from '../PathInfo/PathInfo';

interface AnalyzeLine {
    content: string;
    paths: PathInfo[]
}
export class AnalyzeFiles {
    private files: FileContentNodeAnalyze[];
    constructor(){
        this.files = [];
    }

    addFile(file : FileContent) {
        const fileNode = file.getFileGraph();
        const signNodes = FileContentNodeAnalyze.addPathToAllGraph(fileNode, file.getPathInfo());
        this.files.push(signNodes);
    }

    getNumberOfFile(){
        return this.files.length;
    }

    analyze() {
        return this.analyzeNodes(this.files);
    }

    getLines(nodes: FileContentNodeAnalyze[]): AnalyzeLine[] {
        const result: AnalyzeLine[] = [];

        nodes.forEach(node => {
            const start = node.getStart();
            const end = node.getEnd();
            if(start !== '') {
                result.push({
                    content: start,
                    paths: node.getPaths()
                });
            }
            result.push(
                ...this.getLines(node.getChildren()),
            );
            if(end !== '') {
                result.push({
                    content: end,
                    paths: node.getPaths()
                });
            }

        });

        return result;
    }

    private analyzeNodes(nodes: FileContentNodeAnalyze[]): FileContentNodeAnalyze[] {

        const grouped: FileContentNodeAnalyze[][] = [];
        nodes.forEach(node => {
            const indexToAdd = grouped.findIndex(group => group[0].isSame(node));
            if(indexToAdd === -1) {
                grouped.push([node]);
            } else {
                grouped[indexToAdd].push(node);
            }
        });

        return this.attachChildren(grouped);
    }

    private analyzeChildrenNodes(nodes: FileContentNodeAnalyze[][]): FileContentNodeAnalyze[] {

        const grouped: FileContentNodeAnalyze[][] = [];

        nodes.forEach(children => {
            children.forEach(childrenNode => {
                const indexToAdd = grouped.findIndex(group => group[0].isSame(childrenNode) && group.every(node => !node.isSameBasePath(childrenNode)));
                if(indexToAdd === -1) {
                    grouped.push([childrenNode]);
                } else {
                    grouped[indexToAdd].push(childrenNode);
                }
            });
        });

        return this.attachChildren(grouped);
    }

    private attachChildren(grouped: FileContentNodeAnalyze[][]) {
        return grouped.map(nodes => {
            const [node, ...otherNodes] = nodes;

            const result = this.analyzeChildrenNodes(nodes.map(node => node.getChildren()));
            const newNode = node.setChildren(result);
            otherNodes.map(otherFile => {
                newNode.addPath(otherFile.getBasePath());
            });
            return newNode;
        });
    }
}
