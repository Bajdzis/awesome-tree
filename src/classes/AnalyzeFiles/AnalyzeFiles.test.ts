import { FileContent } from '../FileContent/FileContent';
import { FileContentNode } from '../FileContentNode/FileContentNode';
import { PathInfo } from '../PathInfo/PathInfo';
import { AnalyzeFiles } from './AnalyzeFiles';


const someContent = 'line1\nline2\nline3';
const invalidContent = 'line1\nline3';

const path1 = new PathInfo('/home/document/file1.txt');
const path2 = new PathInfo('/home/document/file2.txt');
const path3 = new PathInfo('/home/document/file3.txt');
const path4 = new PathInfo('/home/document/file4.txt');

const file1 = new FileContent(path1, someContent);
const file2 = new FileContent(path2, someContent);
const file3 = new FileContent(path3, invalidContent);
const file4 = new FileContent(path4, someContent);

describe('AnalyzeFiles', () => {

    it('should return analyze compare', () => {

        const analyzer = new AnalyzeFiles();
        analyzer.addFile(file1);
        analyzer.addFile(file2);
        analyzer.addFile(file3);
        analyzer.addFile(file4);

        expect(analyzer.analyze()).toEqual({});
    });


});
