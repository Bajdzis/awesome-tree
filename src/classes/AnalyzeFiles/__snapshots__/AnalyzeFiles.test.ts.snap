// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyzeFiles should return analyze compare 1`] = `
Array [
  FileContentNodeAnalyze {
    "body": Array [
      FileContentNodeAnalyze {
        "body": Array [],
        "end": "",
        "paths": Array [
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file1.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file2.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file3.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file4.txt",
            "type": "file",
          },
        ],
        "start": "line1
",
      },
      FileContentNodeAnalyze {
        "body": Array [],
        "end": "",
        "paths": Array [
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file1.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file2.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file4.txt",
            "type": "file",
          },
        ],
        "start": "line2
",
      },
      FileContentNodeAnalyze {
        "body": Array [],
        "end": "",
        "paths": Array [
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file1.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file2.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file3.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file4.txt",
            "type": "file",
          },
        ],
        "start": "line3
",
      },
      FileContentNodeAnalyze {
        "body": Array [],
        "end": "",
        "paths": Array [
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file1.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file2.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file3.txt",
            "type": "file",
          },
          Object {
            "name": "PathInfo",
            "parts": Array [
              WordsInfo {
                "parts": Array [
                  "home",
                ],
                "words": "home",
              },
              WordsInfo {
                "parts": Array [
                  "document",
                ],
                "words": "document",
              },
              WordsInfo {
                "parts": Array [
                  "file",
                ],
                "words": "file",
              },
              WordsInfo {
                "parts": Array [
                  "txt",
                ],
                "words": "txt",
              },
            ],
            "path": "/home/document/file4.txt",
            "type": "file",
          },
        ],
        "start": "",
      },
    ],
    "end": "",
    "paths": Array [
      Object {
        "name": "PathInfo",
        "parts": Array [
          WordsInfo {
            "parts": Array [
              "home",
            ],
            "words": "home",
          },
          WordsInfo {
            "parts": Array [
              "document",
            ],
            "words": "document",
          },
          WordsInfo {
            "parts": Array [
              "file",
            ],
            "words": "file",
          },
          WordsInfo {
            "parts": Array [
              "txt",
            ],
            "words": "txt",
          },
        ],
        "path": "/home/document/file1.txt",
        "type": "file",
      },
      Object {
        "name": "PathInfo",
        "parts": Array [
          WordsInfo {
            "parts": Array [
              "home",
            ],
            "words": "home",
          },
          WordsInfo {
            "parts": Array [
              "document",
            ],
            "words": "document",
          },
          WordsInfo {
            "parts": Array [
              "file",
            ],
            "words": "file",
          },
          WordsInfo {
            "parts": Array [
              "txt",
            ],
            "words": "txt",
          },
        ],
        "path": "/home/document/file2.txt",
        "type": "file",
      },
      Object {
        "name": "PathInfo",
        "parts": Array [
          WordsInfo {
            "parts": Array [
              "home",
            ],
            "words": "home",
          },
          WordsInfo {
            "parts": Array [
              "document",
            ],
            "words": "document",
          },
          WordsInfo {
            "parts": Array [
              "file",
            ],
            "words": "file",
          },
          WordsInfo {
            "parts": Array [
              "txt",
            ],
            "words": "txt",
          },
        ],
        "path": "/home/document/file3.txt",
        "type": "file",
      },
      Object {
        "name": "PathInfo",
        "parts": Array [
          WordsInfo {
            "parts": Array [
              "home",
            ],
            "words": "home",
          },
          WordsInfo {
            "parts": Array [
              "document",
            ],
            "words": "document",
          },
          WordsInfo {
            "parts": Array [
              "file",
            ],
            "words": "file",
          },
          WordsInfo {
            "parts": Array [
              "txt",
            ],
            "words": "txt",
          },
        ],
        "path": "/home/document/file4.txt",
        "type": "file",
      },
    ],
    "start": "",
  },
]
`;
